src/solvers/dg.jl

```diff
@@ -0,0 +1,482 @@
+    #     is probably the best option since everything will be handed over to
+    #     Chris Elrod, one of the best performance software engineers for Julia.
+    PtrArray(pointer(u_ode),
+             (StaticInt(nvariables(equations)), ntuple(_ -> StaticInt(nnodes(dg)), ndims(mesh))..., nelements(dg, cache)))
+            #  (nvariables(equations), ntuple(_ -> nnodes(dg), ndims(mesh))..., nelements(dg, cache)))
+  else
+    # The following version is reasonably fast and allows us to `resize!(u_ode, ...)`.
+    unsafe_wrap(Array{eltype(u_ode), ndims(mesh)+2}, pointer(u_ode),
+                (nvariables(equations), ntuple(_ -> nnodes(dg), ndims(mesh))..., nelements(dg, cache)))
+  end
+end
+
+# General fallback
+@inline function wrap_array(u_ode::AbstractVector, mesh::AbstractMesh, equations, dg::DG, cache)
+  wrap_array_native(u_ode, mesh, equations, dg, cache)
+end
+
+# Like `wrap_array`, but guarantees to return a plain `Array`, which can be better
+# for interfacing with external C libraries (MPI, HDF5, visualization),
+# writing solution files etc.
+@inline function wrap_array_native(u_ode::AbstractVector, mesh::AbstractMesh, equations, dg::DG, cache)
+  @boundscheck begin
+    @assert length(u_ode) == nvariables(equations) * nnodes(dg)^ndims(mesh) * nelements(dg, cache)
+  end
+  unsafe_wrap(Array{eltype(u_ode), ndims(mesh)+2}, pointer(u_ode),
+              (nvariables(equations), ntuple(_ -> nnodes(dg), ndims(mesh))..., nelements(dg, cache)))
+end
+
+
+function compute_coefficients!(u, func, t, mesh::AbstractMesh{1}, equations, dg::DG, cache)
+
+  @threaded for element in eachelement(dg, cache)
+    for i in eachnode(dg)
+      x_node = get_node_coords(cache.elements.node_coordinates, equations, dg, i, element)
+      u_node = func(x_node, t, equations)
+      set_node_vars!(u, u_node, equations, dg, i, element)
+    end
+  end
+end
+
+function compute_coefficients!(u, func, t, mesh::AbstractMesh{2}, equations, dg::DG, cache)
+
+  @threaded for element in eachelement(dg, cache)
+    for j in eachnode(dg), i in eachnode(dg)
+      x_node = get_node_coords(cache.elements.node_coordinates, equations, dg, i, j, element)
+      u_node = func(x_node, t, equations)
+      set_node_vars!(u, u_node, equations, dg, i, j, element)
+    end
+  end
+end
+
+function compute_coefficients!(u, func, t, mesh::AbstractMesh{3}, equations, dg::DG, cache)
+
+  @threaded for element in eachelement(dg, cache)
+    for k in eachnode(dg), j in eachnode(dg), i in eachnode(dg)
+      x_node = get_node_coords(cache.elements.node_coordinates, equations, dg, i, j, k, element)
+      u_node = func(x_node, t, equations)
+      set_node_vars!(u, u_node, equations, dg, i, j, k, element)
+    end
+  end
+end
+
+
+# Discretizations specific to each mesh type of Trixi
+# If some functionality is shared by multiple combinations of meshes/solvers,
+# it is defined in the directory of the most basic mesh and solver type.
```

Comment:
I like this convention and the fact that it's now been "officially" written down somewhere.

I always had the impression that, to have everything neat and tidy, we would have to reorganize a lot of code to have, e.g., all curved stuff in an extra file. This convention definitely makes things a lot easier. No need to think about where code belongs (is it specific to curved? Is it specific to DGSEM?) and no need to move stuff every time an existing functionality is added to another solver.